File: tests/src/test_dataset.cpp
#include <gtest/gtest.h>

#include <filesystem>
#include <fstream>
#include <map>
#include <memory>
#include <optional>
#include <vector>

#include "dataset.hpp"
#include "subject.hpp"

class DatasetTest : public ::testing::Test {
 protected:
  std::filesystem::path test_dir;
  std::filesystem::path participants_file;
  std::filesystem::path participants_sidecar_file;
  std::unique_ptr<Dataset> dataset;

  void SetUp() override {
    // Create a temporary test directory
    test_dir = std::filesystem::temp_directory_path() / "dataset_test";
    std::filesystem::create_directory(test_dir);

    // Set up file paths
    participants_file = test_dir / "participants.tsv";
    participants_sidecar_file = test_dir / "participants.json";
    std::filesystem::create_directories(test_dir / "sub-01");
    std::filesystem::create_directories(test_dir / "sub-02");

    // Create participants table file
    std::ofstream participants_table(participants_file);
    participants_table << "participant_id\tname\n";
    participants_table << "sub-01\tJohn\n";
    participants_table << "sub-02\tAlice\n";
    participants_table.close();

    // Create participants sidecar file
    std::ofstream participants_sidecar(participants_sidecar_file);
    participants_sidecar << "{\n";
    participants_sidecar << "\t\"name\": {\n";
    participants_sidecar << "\t\t\"Description\": \"Name of participant\"\n";
    participants_sidecar << "\t}\n";
    participants_sidecar << "}\n";
    participants_sidecar.close();

    // Create the Dataset object
    dataset = std::make_unique<Dataset>(test_dir, true);
  }

  void TearDown() override {
    // Remove the temporary test directory and files
    std::filesystem::remove_all(test_dir);
  }
};

TEST_F(DatasetTest, AddSubject_ValidArguments_ReturnsSubject) {
  std::map<std::string, std::string> args = {{"participant_id", "sub-03"},
                                             {"name", "Bob"}};

  std::optional<Subject> subject = dataset->add_subject(args);

  EXPECT_TRUE(subject.has_value());
  EXPECT_EQ((*subject)["participant_id"], "sub-03");
  EXPECT_EQ((*subject)["name"], "Bob");
}

TEST_F(DatasetTest, AddSubject_TooManyArguments_ReturnsNullopt) {
  std::map<std::string, std::string> args = {
      {"participant_id", "sub-03"},
      {"name", "Bob"},
      {"age", "25"}  // Extra argument
  };

  EXPECT_DEATH({ std::optional<Subject> subject = dataset->add_subject(args); },
               "Assertion failed");
}

TEST_F(DatasetTest, GetSubject_ExistingSubject_ReturnsSubject) {
  std::optional<Subject> subject = dataset->get_subject<Subject>(1);

  EXPECT_TRUE(subject.has_value());
  EXPECT_EQ((*subject)["participant_id"], "sub-01");
  EXPECT_EQ((*subject)["name"], "John");
}

TEST_F(DatasetTest, GetSubject_NonexistentSubject_ReturnsNullopt) {
  std::optional<Subject> subject = dataset->get_subject<Subject>(3);

  EXPECT_FALSE(subject.has_value());
}

TEST_F(DatasetTest, IsSubject_ExistingSubject_ReturnsTrue) {
  bool result = dataset->is_subject(2);

  EXPECT_TRUE(result);
}

TEST_F(DatasetTest, IsSubject_NonexistentSubject_ReturnsFalse) {
  bool result = dataset->is_subject(3);

  EXPECT_FALSE(result);
}

TEST_F(DatasetTest, GetExistingSubjects_ReturnsSubjectIds) {
  std::vector<int> subject_ids = dataset->get_existing_subjects();

  EXPECT_EQ(subject_ids.size(), 2);
  EXPECT_EQ(subject_ids[0], 1);
  EXPECT_EQ(subject_ids[1], 2);
}

File: tests/test_dataset.py
import pytest
import tempfile
import shutil
from pathlib import Path
from libbids import Dataset


class TestDataset:
    @pytest.fixture(autouse=True)
    def setup(self):
        self.test_dir = Path(tempfile.mkdtemp())
        self.participants_file = self.test_dir / "participants.tsv"
        self.participants_sidecar_file = self.test_dir / "participants.json"
        (self.test_dir / "sub-01").mkdir()
        (self.test_dir / "sub-02").mkdir()

        # Create participants table file

        with open(self.participants_file, "w") as file:
            file.write("participant_id\tname\n")
            file.write("sub-01\tJohn\n")
            file.write("sub-02\tAlice\n")

        # Create participants sidecar file
        with open(self.participants_sidecar_file, "w") as file:
            file.write("{\n")
            file.write('\t"name": {\n')
            file.write('\t\t"Description": "Name of participant"\n')
            file.write("\t}\n")
            file.write("}\n")

        self.dataset = Dataset(self.test_dir, True)

        yield

        # Clean up the temporary test directory and files
        shutil.rmtree(self.test_dir)

    def test_add_subject_valid_arguments_returns_subject(self):
        args = {"participant_id": "sub-03", "name": "Bob"}

        subject = self.dataset.add_subject(args)

        assert subject is not None
        assert subject["participant_id"] == "sub-03"
        assert subject["name"] == "Bob"

    def test_add_subject_too_many_arguments_returns_nullopt(self):
        args = {"participant_id": "sub-03", "name": "Bob", "age": "25"}

        with pytest.raises(Exception):
            self.dataset.add_subject(args)

    def test_get_subject_existing_subject_returns_subject(self):
        subject = self.dataset.get_subject(1)

        assert subject is not None
        assert subject["participant_id"] == "sub-01"
        assert subject["name"] == "John"

    def test_get_subject_nonexistent_subject_returns_nullopt(self):
        subject = self.dataset.get_subject(3)

        assert subject is None

    def test_is_subject_existing_subject_returns_true(self):
        result = self.dataset.is_subject(2)

        assert result is True

    def test_is_subject_nonexistent_subject_returns_false(self):
        result = self.dataset.is_subject(3)

        assert result is False

    def test_get_subjects_returns_subject_ids(self):
        subject_ids = self.dataset.get_subjects()

        assert len(subject_ids) == 2
        assert subject_ids[0] == 1
        assert subject_ids[1] == 2
